### PR TITLE
Improve UpdateChecker

### PR DESCRIPTION
**Pull request by apc999**
### What changes are proposed in this pull request?

Improve UpdateChecker

### Why are the changes needed?

  1. remove irrelevant master features 
  2. turn on fuse update check

### CODE REVIEW

File: dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java
Comment (apc999) [inline dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java:34]: @dbw9580  is this purposely named as `fuse.update.check.enabled` rather than `alluxio.fuse.update.check.enabled` ?

File: dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java
Comment (apc999) [inline dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java:44]: @dbw9580  will this rename create issue?

File: dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java
Comment (dbw9580) [reply dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java:34]: The rationale for not defining it in `PropertyKey.java` and not starting with `alluxio.` is that this is purely a command line option, and it should not be made accessible in `alluxio-site.properties`, etc. I named it without the alluxio prefix and put it here so that the usage should be limited to fuse CLI facilities, hopefully.

File: dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java
Comment (dbw9580) [reply dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java:44]: Renaming this should not cause breakage in functionality, but I purposely did not follow the naming convention to indicate that this a property should be limited to the fuse component, and should not be used elsewhere.

**Comment by LuQQiu** [comment]
@Xenorith Add rico as the gate keeper before merging this PR

**Comment by LuQQiu** [comment]
Discussed with @Xenorith that we will add a ProcessType: FUSE to the checker
and generate the example result to him before merging this PR

File: dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java
Comment (LuQQiu) [reply dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java:34]: The CLI can be removed and not allow user to disable it from this approach.
do we consider using 
```
dev/scripts/src/alluxio.org/build/profiles.yml:    mvnArgs: "-Dupdate.check.enabled=false"
docs/_data/generated/en/master-configuration.yml:alluxio.master.update.check.enabled:
docs/_data/generated/master-configuration.csv:alluxio.master.update.check.enabled,"true"
dora/core/common/src/main/java-templates/alluxio/ProjectConstants.java:  public static final String UPDATE_CHECK_ENABLED = "${update.check.enabled}";
dora/core/common/src/main/java/alluxio/conf/PropertyKey.java:        "alluxio.master.update.check.enabled";
dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java:      PropertyKey.Builder.booleanBuilder("fuse.update.check.enabled")
pom.xml:    <update.check.enabled>true</update.check.enabled>
```
the update check maven arguments way for this fuse update check disable during internal development and testing?

Comment (Xenorith) [reply dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java:34]: i am also worried about hardcoding this. there needs to be a way to differentiate whether or not the update checks should run by default. the current mechanism is to set `-Dupdate.check.enabled=false` when building the code for specific purposes; see https://github.com/Alluxio/alluxio/blob/672562bc96b3bb575ad66e3d8e765ca2794b089e/dev/scripts/src/alluxio.org/build/profiles.yml#L127C1-L127C1

i recommend reusing the same mechanism and set `.setDefaultValue(Boolean.parseBoolean(ProjectConstants.UPDATE_CHECK_ENABLED))` for this property

Comment (apc999) [reply dora/integration/fuse/src/main/java/alluxio/fuse/options/FuseOptions.java]: followed what @Xenorith suggested

File: dora/integration/fuse/src/main/java/alluxio/fuse/meta/FuseUpdateChecker.java
Comment (apc999) [inline dora/integration/fuse/src/main/java/alluxio/fuse/meta/FuseUpdateChecker.java:60]: @Xenorith 👆

File: dora/core/common/src/main/java/alluxio/util/UpdateCheckUtils.java
Comment (Xenorith) [inline dora/core/common/src/main/java/alluxio/util/UpdateCheckUtils.java:122]: is this removal on purpose? i do not see these user agent features readded for the master side update check

**Review by Xenorith — COMMENTED** [review]
previously the update checker on master has a "scale" of 1 update check per cluster deployment. what's the approximate scale for the fuse update checker? we would need to be more careful if the scale suddenly is 100-1000 times larger.

File: dora/core/common/src/main/java/alluxio/util/UpdateCheckUtils.java
Comment (apc999) [reply dora/core/common/src/main/java/alluxio/util/UpdateCheckUtils.java:122]: yes, this is intentional. original `addUserAgentFeatures` method is only covering master features and none of them is relevant after 300. Now, if we need any feature related info, make it in their respective UpdateChecker

**Comment by apc999** [comment]
@LuQQiu  @Xenorith  PTAL

**Review by Xenorith — COMMENTED** [review]
these changes look good to me, but i think it will be a good idea to update the endpoint's version to reflect the set of new changes. https://github.com/Alluxio/alluxio/blob/main/dora/core/common/src/main/java/alluxio/check/UpdateCheck.java#L83 is using a path with `/v0` and i would recommend to change it to `/v1`. but i will also have to make the corresponding changes on the receiving side for this to properly work, so please hold off on merging until it is ready

**Comment by Xenorith** [comment]
@LuQQiu can i get a (fully populated) example of a http request to compare the server side with?

**Comment by LuQQiu** [comment]
@Xenorith @Zijian-Zhu 
the others should be similar to before, but userAgent string is different
```
Alluxio/304-SNAPSHOT (59eea89a-8986-49d2-93b3-528d5e6243ef; OS:Mac OS X; ec2; localAlluxioMetadataCache; localKernelDataCache; UnderlyingFileSystem:s3; processType:FUSE)
```
User agent string contains ID, agent environments, processType similar to original user agent
```
  @VisibleForTesting
  public static String getUserAgentString(String id, CommonUtils.ProcessType processType,
      List<String> additionalInfo) {
    List<String> info = new ArrayList<>();
    info.add(id);
    addUserAgentEnvironments(info);
    info.addAll(additionalInfo);
    info.add(String.format("processType:%s", processType.toString()));
    return String.format("Alluxio/%s (%s)", ProjectConstants.VERSION,
        Joiner.on(USER_AGENT_SEPARATOR + " ").skipNulls().join(info));
  }
```
```
    if (fuseOptions.getFileSystemOptions().isDataCacheEnabled()) {
      mFuseInfo.add(LOCAL_ALLUXIO_DATA_CACHE);
    }
    if (fuseOptions.getFileSystemOptions().isMetadataCacheEnabled()) {
      mFuseInfo.add(LOCAL_ALLUXIO_METADATA_CACHE);
    }
    if (!fuseOptions.getFuseMountOptions().contains("direct_io")) {
      mFuseInfo.add(LOCAL_KERNEL_DATA_CACHE);
    }
```
UnderFileSystemType: "alluxio" or "local" or ufs type (e.g. "s3", "hdfs"
```
    mFuseInfo.add(String.format("UnderlyingFileSystem:%s", getUnderlyingFileSystem(fuseOptions)));
    if (!fuseOptions.getFileSystemOptions().getUfsFileSystemOptions().isPresent()) {
      return ALLUXIO_FS;
    }
    String ufsAddress = fuseOptions.getFileSystemOptions()
        .getUfsFileSystemOptions().get().getUfsAddress();
    if (URIUtils.isLocalFilesystem(ufsAddress)) {
      return LOCAL_FS;
    }
    String[] components = ufsAddress.split(":");
    if (components.length < 2) {
      return "unknown";
    }
```

**Comment by Zijian-Zhu** [comment]
the update for receiving side is done. Please update the endpoint from path `/v0`to `/v1`.

**Comment by Xenorith** [comment]
guessing the integration test failure is real:
```
2023-11-01T21:29:23.3531640Z 21:29:23.351 [ERROR] Tests run: 8, Failures: 2, Errors: 0, Skipped: 1, Time elapsed: 1.057 s <<< FAILURE! - in alluxio.fuse.meta.FuseUpdateCheckerTest
2023-11-01T21:29:23.3533855Z 21:29:23.352 [ERROR] alluxio.fuse.meta.FuseUpdateCheckerTest.UnderFileSystemAlluxio  Time elapsed: 0.01 s  <<< FAILURE!
2023-11-01T21:29:23.3534882Z java.lang.AssertionError
2023-11-01T21:29:23.3535407Z 	at org.junit.Assert.fail(Assert.java:87)
2023-11-01T21:29:23.3536037Z 	at org.junit.Assert.assertTrue(Assert.java:42)
2023-11-01T21:29:23.3536961Z 	at org.junit.Assert.assertTrue(Assert.java:53)
2023-11-01T21:29:23.3538059Z 	at alluxio.fuse.meta.FuseUpdateCheckerTest.UnderFileSystemAlluxio(FuseUpdateCheckerTest.java:37)
2023-11-01T21:29:23.3539200Z 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-11-01T21:29:23.3540370Z 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2023-11-01T21:29:23.3541553Z 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-11-01T21:29:23.3542531Z 	at java.lang.reflect.Method.invoke(Method.java:498)
2023-11-01T21:29:23.3544985Z 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
2023-11-01T21:29:23.3546584Z 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2023-11-01T21:29:23.3548070Z 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
2023-11-01T21:29:23.3549438Z 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2023-11-01T21:29:23.3550535Z 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
2023-11-01T21:29:23.3551606Z 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
2023-11-01T21:29:23.3552686Z 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
2023-11-01T21:29:23.3553776Z 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
2023-11-01T21:29:23.3555141Z 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
2023-11-01T21:29:23.3556150Z 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
2023-11-01T21:29:23.3557021Z 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
2023-11-01T21:29:23.3557949Z 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
2023-11-01T21:29:23.3558872Z 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
2023-11-01T21:29:23.3559788Z 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
2023-11-01T21:29:23.3560698Z 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
2023-11-01T21:29:23.3561642Z 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
2023-11-01T21:29:23.3562656Z 	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:316)
2023-11-01T21:29:23.3563929Z 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:240)
2023-11-01T21:29:23.3565249Z 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:214)
2023-11-01T21:29:23.3566685Z 	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:155)
2023-11-01T21:29:23.3568114Z 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
2023-11-01T21:29:23.3570466Z 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
2023-11-01T21:29:23.3571867Z 	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
2023-11-01T21:29:23.3573303Z 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
```
@LuQQiu do you know what could be wrong?

**Review by Xenorith — APPROVED** [review]

**Comment by Xenorith** [comment]
alluxio-bot, merge this please

**Pull request merged**
